### PR TITLE
fix: callback and callback group struct

### PR DESCRIPTION
**Pull request by h-suzuki-isp**
## Description

When creating `CallbackGroup` data from trace data in `LttngInfo`, include those that are not linked to executor. In this case, since the `callback_group_type` cannot be obtained for those not linked to an executor, assign `UNDEFINED`.

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines](https://github.com/tier4/caret/blob/main/.github/CONTRIBUTING.md).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] (Optional) The PR has been properly tested with CARET_report verification.
- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: src/caret_analyze/architecture/architecture.py [hunk 1]
```diff
@@ -77,9 +77,11 @@ def get_node(self, node_name: str) -> NodeStructValue:
...
         return Util.find_one(lambda x: x.executor_name == executor_name, self.executors)
 
-    def get_callback_group(self, callback_group_name: str) -> CallbackGroupStructValue:
-        return Util.find_one(lambda x: x.callback_group_name == callback_group_name,
-                             self.callback_groups)
+    def get_callback_group(self) -> tuple[CallbackGroupStructValue, ...]:
+        cbg: list[CallbackGroupStructValue] = []
+        for node in self.nodes:
+            cbg += node.callback_groups
+        return tuple(set(cbg))
```
Comment (ymski) [inline src/caret_analyze/architecture/architecture.py:84]: get_callback_group() is a function that gets the CallbackGroupStruct corresponding to a given callback_group_name.
It does not get all callback groups.

Comment (h-suzuki-isp) [reply src/caret_analyze/architecture/architecture.py]: I seem to have misunderstood something.
Fixed in https://github.com/tier4/caret_analyze/pull/511/commits/ce7a66f001a7fd44eb8fb3d6f4c7cdf0c219d764.

**Review by isp-uetsuki — CHANGES_REQUESTED** [review]
- There are some changes like `if xxx is None` --> `if not xxx` and `if xxx is not None` --> `if xxx`. I suggest reverting these changes because original codes are slightly better. (Original codes represent directly what we want.)
- There are some changes to avoid None of `callback_group` and `variable_passing`. I suggest reverting these changes because they are not necessary.

File: src/caret_analyze/architecture/architecture.py [hunk 2]
```diff
@@ -83,7 +83,10 @@ def get_callback_group(self, callback_group_name: str) -> CallbackGroupStructVal
 
     @property
     def callback_groups(self) -> tuple[CallbackGroupStructValue, ...]:
-        return tuple(Util.flatten(_.callback_groups for _ in self.executors))
+        cbg: list[CallbackGroupStructValue] = []
+        for node in self.nodes:
+            cbg += node.callback_groups
+        return tuple(set(cbg))
```
[the remark this answers is not in the document]
Comment (h-suzuki-isp) [reply src/caret_analyze/architecture/architecture.py]: Fixed in https://github.com/tier4/caret_analyze/pull/511/commits/b68f2911e6abbb3b859eec95257d2aa2d9e5e5f9.

File: src/test/infra/lttng/test_lttng_info.py [hunk 3]
```diff
@@ -646,6 +646,105 @@ def test_get_callback_groups_info(self, mocker):
...
+        cbg = TracePointData(pd.DataFrame.from_dict(
+            [
+                {
+                    'callback_group_id': pd.NA,
+                    'callback_group_addr': cbg_addr,
+                    'executor_addr': pd.NA,
+                    'group_type_name': pd.NA,
+                }
+            ]
+        ))
```
[the remark this answers is not in the document]
Comment (h-suzuki-isp) [reply src/test/infra/lttng/test_lttng_info.py]: Fixed in https://github.com/tier4/caret_analyze/pull/511/commits/4ca184504ccc53438496fa43a835d7e57c22540d.

File: src/caret_analyze/infra/lttng/lttng_info.py [hunk 4]
```diff
@@ -640,15 +641,33 @@ def _get_callback_groups(
                 callback_ids = tuple(group_df['callback_id'].values)
                 callback_ids = tuple(Util.filter_items(self._is_user_made_callback, callback_ids))
 
+                # For the case where the callback_group is not linked to the executor
+                if row['executor_addr'] is not pd.NA:
+                    executor_addr = row['executor_addr']
+                else:
+                    executor_addr = dummy_executor_addr
+                    dummy_executor_addr -= 1
```
[the remark this answers is not in the document]
Comment (h-suzuki-isp) [reply src/caret_analyze/infra/lttng/lttng_info.py]: I have confirmed that executor_addr does not need to be unique.
Fixed in https://github.com/tier4/caret_analyze/pull/511/commits/f4c5c7798c5d221403d9ac48b853a5853a449d1e.

File: src/caret_analyze/infra/lttng/value_objects/callback_group.py
[the remark this answers is not in the document]
Comment (h-suzuki-isp) [reply src/caret_analyze/infra/lttng/value_objects/callback_group.py:1]: Fixed in https://github.com/tier4/caret_analyze/pull/511/commits/f58dc0c80cdc77a47a3f37efff3b1babd7637013.

**Comment by h-suzuki-isp** [comment]
@isp-uetsuki 
Thank you for your review.

> There are some changes like if xxx is None --> if not xxx and if xxx is not None --> if xxx. I suggest reverting these changes because original codes are slightly better. (Original codes represent directly what we want.)
There are some changes to avoid None of callback_group and variable_passing. I suggest reverting these changes because they are not necessary.

I fixed in https://github.com/tier4/caret_analyze/pull/511/commits/f1622e1f29478ae221e382d3eef23c47e0853509.

**Review by isp-uetsuki — APPROVED** [review]
LGTM

**Review by miyakoshi-dev — APPROVED** [review]
LGTM
[compare] ========== Result ==========
[compare] 1/3 file existence matches
[compare] OK. file existence matches
[compare] 2/3 yaml files
[compare] OK. yaml matches
[compare] 3/3 PNG files
[compare] OK. PNG matches
[compare] ====== ALL OK ==============

**Pull request merged**